sulong/projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/intrinsics/multithreading/LLVMPThreadIntrinsics.java

```diff
@@ -0,0 +1,847 @@
+import com.oracle.truffle.llvm.runtime.interop.access.LLVMInteropType;
+import com.oracle.truffle.llvm.runtime.memory.LLVMStack;
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMExpressionNode;
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMLoadNode;
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMStoreNode;
+import com.oracle.truffle.llvm.runtime.pointer.LLVMNativePointer;
+import com.oracle.truffle.llvm.runtime.pointer.LLVMPointer;
+import com.oracle.truffle.llvm.runtime.types.FunctionType;
+import com.oracle.truffle.llvm.runtime.types.PointerType;
+import com.oracle.truffle.llvm.runtime.types.Type;
+
+public class LLVMPThreadIntrinsics {
+    public static PrintWriter debugOut = null;
+
+    @CompilerDirectives.TruffleBoundary
+    private static void printDebug(String str) {
+        // create debug output tool
+        if (debugOut == null) {
+            CompilerDirectives.transferToInterpreterAndInvalidate();
+            try {
+                debugOut = new PrintWriter(new FileWriter("/home/florian/debug out"));
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        }
+        debugOut.println(str);
+        debugOut.flush();
+    }
+
+    @NodeChild(type = LLVMExpressionNode.class)
+    @NodeChild(type = LLVMExpressionNode.class)
+    @NodeChild(type = LLVMExpressionNode.class)
+    @NodeChild(type = LLVMExpressionNode.class)
+    public abstract static class LLVMPThreadCreate extends LLVMBuiltin {
+        @Child
+        LLVMStoreNode store = null;
+
+        @Specialization
+        protected int doIntrinsic(VirtualFrame frame, Object thread, Object attr, Object startRoutine, Object arg) {
+            // create store node
+            if (store == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                store = LLVMLanguage.getLLVMContextReference().get().getNodeFactory().createStoreNode(LLVMInteropType.ValueKind.I64);
+            }
+
+            // create thread for execution of function
+            Thread t = LLVMLanguage.getLLVMContextReference().get().getEnv().createThread(() -> {
+                CompilerDirectives.transferToInterpreter();
+                // RootCallTarget callTarget = Truffle.getRuntime().createCallTarget(new RunNewThreadNode(getLLVMLanguage()));
+                RootCallTarget callTarget = Truffle.getRuntime().createCallTarget(new RunNewThreadNode(LLVMLanguage.getLanguage()));
+
+                // pthread_exit throws a control flow exception to stop the thread
+                try {
+                    callTarget.call(startRoutine, arg);
+                } catch (ControlFlowException e) {
+
+                }
+            });
+            // store cur id in thread var
+            store.executeWithTarget(thread, t.getId());
+
+            // store thread with thread id in context
+            LLVMLanguage.getLLVMContextReference().get().threadStorage.put(t.getId(), t);
```

Comment:
Reading from or writing to any collection class from `java.util.**` must only happen in methods that are annotated with `@TruffleBoundary` to prevent the Inliner / Partial Evaluater from going crazy.

Reply:
bekam den fehler:
/home/florian/my-sulong-dev/graal/sulong/projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/intrinsics/multithreading/LLVMPThreadThreadIntrinsics.java:171: error: Method doIntrinsic cannot be annotated with @TruffleBoundary and have a parameter of type VirtualFrame

soll ich also die lese / schreib zugriffe auf die collections auf andere methoden (also nicht direkt in den doIntrinsic) auslagern? oder was schlägst du vor?

Reply:
ja, Zugriff auf Collections sollte in eigenen Methoden passieren, die mit `@TruffleBoundary` annotiert sind.